{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3bc163ea",
   "metadata": {},
   "source": [
    "### Workflow from genomes to families tables\n",
    "* Setup dir structure\n",
    "* Cluster ORFs and select reprseqs\n",
    "* Create HMM profiles with uniclust\n",
    "* Create db from set of profiles\n",
    "* Run all vs all comparison of profiles\n",
    "* Collect resulst of all-vs-all into table\n",
    "\n",
    "This notebook does **not** contain running profiles on HHsuite-formated databases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c821e1f2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "3b8529ac",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/ directory already set up\n",
      "/tmp/repr-proteins/ directory already set up\n",
      "/tmp/mmseqs/ directory already set up\n",
      "/tmp/all-by-all/ directory already set up\n",
      "/tmp/all-by-all/individual-seqs/ directory already set up\n",
      "/tmp/parse/ directory already set up\n",
      "/tmp/prot-families/ directory already set up\n",
      "/tmp/prot-families/pair_table_chunks/ directory already set up\n",
      "/input/ directory already set up\n",
      "/input/phanotate/ directory already set up\n",
      "/input/coding-seqs/ directory already set up\n",
      "/output/ directory already set up\n",
      "/output/prot-families/ directory already set up\n",
      "/output/prot-families/representative/ directory already set up\n",
      "/output/prot-families/all-by-all/ directory already set up\n",
      "//output/prot-families/functional/ directory already set up\n",
      "//output/prot-families/families/ directory already set up\n",
      "/output/prot-families/all-by-all/hhblits/ directory already set up\n",
      "/output/prot-families/all-by-all/mmseqs/ directory already set up\n",
      "/intermediate/ directory already set up\n",
      "/intermediate/prot-families/ directory already set up\n",
      "/intermediate/prot-families/profiles/ directory already set up\n",
      "/intermediate/prot-families/profiles/hhblits/ directory already set up\n",
      "/intermediate/prot-families/functional/ directory already set up\n",
      "/intermediate/prot-families/functional/hhrs/ directory already set up\n",
      "/intermediate/prot-families/profiles/mmseqs/ directory already set up\n",
      "/intermediate/prot-families/all-by-all/ directory already set up\n",
      "/intermediate/prot-families/all-by-all/hhblits/ directory already set up\n",
      "/intermediate/prot-families/all-by-all/mmseqs/ directory already set up\n",
      "/intermediate/prot-families/db/ directory already set up\n",
      "/intermediate/prot-families/db/hhblits/ directory already set up\n",
      "/intermediate/prot-families/db/mmseqs/ directory already set up\n",
      "/log/ directory already set up\n"
     ]
    }
   ],
   "source": [
    "### setup ###\n",
    "\n",
    "### imports\n",
    "\n",
    "import random\n",
    "import sys\n",
    "pipeline_env_path = '/Users/kszczepaniak/Code/phage-pipeline-env'\n",
    "lib_pp_path =       '/Users/kszczepaniak/Code/phage-pipeline-env/phage-pipeline'\n",
    "sys.path.append(lib_pp_path)\n",
    "import submodules.pdf.domain_split as ds\n",
    "from lib_phage.clustering import cluster_proteins\n",
    "from lib_phage.utils import setup_dir_tree, fetch_and_rename_protein_ids, build_hhr_table\n",
    "from lib_phage.utils import process_phanotate_output, create_reprseq_profile_from_clustering\n",
    "from lib_phage.utils import create_bash_script_to_parse_hhr_results, run_parsing_with_bash\n",
    "from lib_phage.utils import concatenate_parsing_results, clean_clustering_partial_data\n",
    "from lib_phage.logs import check_input_repr_prot_selection, validate_output_repr_prot_selection\n",
    "from lib_phage.logs import check_input_all_vs_all_HMM, save_params_hhblits, validate_output_hhblits\n",
    "from lib_phage.logs import validate_create_db, validate_search_all_vs_all, validate_input_ECF, validate_output_ECF\n",
    "from lib_phage.prot_compare import save_individual_seqs, run_hhblits, build_hh_db, run_all_vs_all\n",
    "from lib_phage.ecf_finder_wrapper import load_and_filter_data, store_scan_results\n",
    "\n",
    "from lib_phage.repr_hits_pairwise import get_prob_cov\n",
    "\n",
    "### run mode\n",
    "run_mode = 'hhblits' # profile creation mode [mmseqs/hhblits]\n",
    "\n",
    "### paths\n",
    "\n",
    "# data sources\n",
    "\n",
    "# cds_all_filepath = '/Users/kszczepaniak/Data/Phage/Rafal_Dropbox/data/input/combined/coding-seqs/cds-aa.fa.gz'\n",
    "phanotate_filepath = '/Users/kszczepaniak/Data/Phage/phage-pipeline-input-dev/seq-test.phanotate.txt'\n",
    "\n",
    "# work dirs\n",
    "work_dir = '/Users/kszczepaniak/Data/Phage/phage-pipeline-workdir-dev/'\n",
    "setup_dir_tree(work_dir)\n",
    "\n",
    "# binaries and libraries\n",
    "\n",
    "mmseqs_binpath  = 'mmseqs'\n",
    "uniref_db_path  = '/Users/kszczepaniak/Data/Phage/db/UniRef30_2020_06/UniRef30_2020_06'\n",
    "hhsuite_bins    = '/Users/kszczepaniak/Tools/hh-suite/build/bin'\n",
    "hhsuite_scripts = '/Users/kszczepaniak/Tools/hh-suite/build/scripts'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1f2a196a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All fasta translated. File compressed.\n"
     ]
    }
   ],
   "source": [
    "### Create protein db from Phanotate data ###\n",
    "\n",
    "# translate Phanotate & compress with bgzip\n",
    "process_phanotate_output(phanotate_filepath=phanotate_filepath, work_dir=work_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "caad14e4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating db... Done!\n",
      "Clustering... Done!\n",
      "Generating a clustering table... Done!\n",
      "Creating MSA... Done!\n",
      "Validation success, log file stored.\n"
     ]
    }
   ],
   "source": [
    "### Get representative proteins ###\n",
    "\n",
    "# set clustering params\n",
    "cluster_params_min_seqid   = 0.3\n",
    "cluster_params_sensitivity = 7\n",
    "cluster_params_coverage    = 0.95\n",
    "\n",
    "# check input files integrity & if this step was already executed:\n",
    "# if it was warn about data overwrite\n",
    "if check_input_repr_prot_selection():\n",
    "\n",
    "    # perform clustering\n",
    "    clustering_filepath, clustering_msa_filepath = cluster_proteins(input_fasta_filepath=work_dir + 'input/coding-seqs/cds-aa.fa.gz',\n",
    "                                           output_dirpath=work_dir + 'output/prot-families/representative',\n",
    "                                           mmseqs_tempdir=work_dir + 'tmp/mmseqs',\n",
    "                                           mmseqs_binpath=mmseqs_binpath,\n",
    "                                           cluster_params_min_seqid=cluster_params_min_seqid,\n",
    "                                           cluster_params_sensitivity=cluster_params_sensitivity,\n",
    "                                           cluster_params_coverage=cluster_params_coverage,\n",
    "                                           verbose=True)\n",
    "\n",
    "    no_repr_prot, name_table_filepath = fetch_and_rename_protein_ids(work_dir, clustering_filepath, \n",
    "                                                                     work_dir + 'input/coding-seqs/cds-aa.fa.gz')\n",
    "\n",
    "    # verify output and save log file\n",
    "    validate_output_repr_prot_selection(work_dir=work_dir,\n",
    "                                        output_dirpath=work_dir + 'output/prot-families/representative',\n",
    "                                        cluster_params_min_seqid=cluster_params_min_seqid,\n",
    "                                        cluster_params_sensitivity=cluster_params_sensitivity,\n",
    "                                        cluster_params_coverage=cluster_params_coverage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0fd9a2ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create profiles from mmseqs clusters ###\n",
    "# profile for singleton in clustering is the sequence itself\n",
    "\n",
    "create_reprseq_profile_from_clustering(clustering_filepath, clustering_msa_filepath,\n",
    "                                       cds_all_filepath = work_dir + 'input/coding-seqs/cds-aa.fa',\n",
    "                                       profile_outdir = work_dir + 'intermediate/prot-families/profiles/mmseqs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b2a3e5a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This will overwrite all data from this step. Proceed? [y/n]y\n",
      "Clearing /Users/kszczepaniak/Data/Phage/phage-pipeline-workdir-dev/intermediate/prot-families/all-by-all...\n",
      "Parameters saved, log file stored.\n"
     ]
    }
   ],
   "source": [
    "### Perform all vs all comparison ###\n",
    "# create profiles for each protein with hhblits #\n",
    "\n",
    "# set create profiles params\n",
    "cpu  = 2 # max number of CPUs to be used in the step\n",
    "n    = 1\n",
    "mact = 0.35\n",
    "p    = 90\n",
    "qid  = 10\n",
    "cov  = 10\n",
    "\n",
    "if run_mode == 'hhblits': # execute this only when creating profiles with hhblits\n",
    "\n",
    "    # validate previous step\n",
    "    if check_input_all_vs_all_HMM(work_dir=work_dir, force=True):\n",
    "\n",
    "        # execute current step\n",
    "        save_individual_seqs(work_dir=work_dir)\n",
    "\n",
    "        run_hhblits(work_dir=work_dir, hhsuite_bins=hhsuite_bins, hhsuite_scripts=hhsuite_scripts, cpu=cpu, \n",
    "                    uniref_db_path=uniref_db_path, n=n, mact=mact, p=p, qid=qid, cov=cov)\n",
    "\n",
    "        # save params to log\n",
    "        save_params_hhblits(work_dir=work_dir, n=n, mact=mact, p=p, qid=qid, cov=cov)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "67884e16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Database already exists. Overwrite? [y/n]y\n",
      "Database cleaned.\n",
      "Concatenated a3m alignments.\n",
      "Created HMM profiles.\n",
      "Created column state (CS) sequence database.\n",
      "DB sorted.\n",
      "DB successfuly created.\n"
     ]
    }
   ],
   "source": [
    "# Build db #\n",
    "\n",
    "if run_mode == 'hhblits': # execute this only when creating profiles with hhblits\n",
    "\n",
    "    # validate previous step\n",
    "    if validate_output_hhblits(work_dir):\n",
    "\n",
    "        # execute current step\n",
    "        build_hh_db(work_dir=work_dir, hhsuite_bins=hhsuite_bins,\n",
    "                    hhsuite_scripts=hhsuite_scripts, verbose=True, run_mode=run_mode)\n",
    "        \n",
    "elif run_mode == 'mmseqs':\n",
    "    # execute current step\n",
    "    build_hh_db(work_dir=work_dir, hhsuite_bins=hhsuite_bins,\n",
    "                hhsuite_scripts=hhsuite_scripts, verbose=True, run_mode=run_mode)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1253b5af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parameters saved, log file stored.\n"
     ]
    }
   ],
   "source": [
    "# search all vs all #\n",
    "\n",
    "# set all vs all search params\n",
    "cpu  = 2 # max number of CPUs to be used in the step\n",
    "n    = 1\n",
    "p    = 90\n",
    "\n",
    "if validate_create_db(work_dir=work_dir):\n",
    "\n",
    "    run_all_vs_all(work_dir=work_dir, hhsuite_bins=hhsuite_bins, \n",
    "                   hhsuite_scripts=hhsuite_scripts, cpu=cpu, n=n, \n",
    "                   p=p, a3m_wildcard='reprseq*a3m', run_mode=run_mode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "75f6e16e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HHblits all-vs-all step complete. Updated status.\n"
     ]
    }
   ],
   "source": [
    "# create results table #\n",
    "\n",
    "# check if previous step complete\n",
    "if validate_search_all_vs_all(work_dir, run_mode=run_mode):\n",
    "    \n",
    "    # create results table\n",
    "    build_hhr_table(work_dir, run_mode=run_mode)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "950b4a4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qname</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>qlength</th>\n",
       "      <th>sname</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>slength</th>\n",
       "      <th>pident</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>eval</th>\n",
       "      <th>prob</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>reprseq00001</td>\n",
       "      <td>1</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>reprseq00001</td>\n",
       "      <td>1</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>246.5</td>\n",
       "      <td>5.300000e-47</td>\n",
       "      <td>100.0</td>\n",
       "      <td>7.700000e-51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>reprseq00001</td>\n",
       "      <td>30</td>\n",
       "      <td>44</td>\n",
       "      <td>86</td>\n",
       "      <td>reprseq14105</td>\n",
       "      <td>27</td>\n",
       "      <td>41</td>\n",
       "      <td>73</td>\n",
       "      <td>20</td>\n",
       "      <td>25.8</td>\n",
       "      <td>9.300000e-01</td>\n",
       "      <td>56.2</td>\n",
       "      <td>1.100000e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>reprseq00002</td>\n",
       "      <td>1</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>reprseq00002</td>\n",
       "      <td>1</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>100</td>\n",
       "      <td>374.6</td>\n",
       "      <td>8.800000e-65</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.300000e-68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>reprseq00002</td>\n",
       "      <td>1</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>reprseq13080</td>\n",
       "      <td>1</td>\n",
       "      <td>125</td>\n",
       "      <td>138</td>\n",
       "      <td>40</td>\n",
       "      <td>364.7</td>\n",
       "      <td>5.900000e-62</td>\n",
       "      <td>100.0</td>\n",
       "      <td>8.200000e-66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>reprseq00002</td>\n",
       "      <td>1</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>reprseq17146</td>\n",
       "      <td>28</td>\n",
       "      <td>158</td>\n",
       "      <td>158</td>\n",
       "      <td>33</td>\n",
       "      <td>367.2</td>\n",
       "      <td>3.400000e-61</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4.600000e-65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382435</th>\n",
       "      <td>reprseq34463</td>\n",
       "      <td>55</td>\n",
       "      <td>76</td>\n",
       "      <td>200</td>\n",
       "      <td>reprseq32916</td>\n",
       "      <td>39</td>\n",
       "      <td>60</td>\n",
       "      <td>113</td>\n",
       "      <td>27</td>\n",
       "      <td>29.7</td>\n",
       "      <td>1.100000e+00</td>\n",
       "      <td>51.5</td>\n",
       "      <td>1.500000e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382436</th>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>1</td>\n",
       "      <td>187</td>\n",
       "      <td>187</td>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>1</td>\n",
       "      <td>187</td>\n",
       "      <td>187</td>\n",
       "      <td>100</td>\n",
       "      <td>463.6</td>\n",
       "      <td>1.400000e-72</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.100000e-76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382437</th>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>9</td>\n",
       "      <td>183</td>\n",
       "      <td>187</td>\n",
       "      <td>reprseq06995</td>\n",
       "      <td>1</td>\n",
       "      <td>181</td>\n",
       "      <td>181</td>\n",
       "      <td>32</td>\n",
       "      <td>432.2</td>\n",
       "      <td>8.900000e-68</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.400000e-71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382438</th>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>6</td>\n",
       "      <td>73</td>\n",
       "      <td>187</td>\n",
       "      <td>reprseq13037</td>\n",
       "      <td>48</td>\n",
       "      <td>109</td>\n",
       "      <td>201</td>\n",
       "      <td>27</td>\n",
       "      <td>34.7</td>\n",
       "      <td>5.900000e-01</td>\n",
       "      <td>63.0</td>\n",
       "      <td>7.000000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382439</th>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>44</td>\n",
       "      <td>73</td>\n",
       "      <td>187</td>\n",
       "      <td>reprseq04060</td>\n",
       "      <td>14</td>\n",
       "      <td>43</td>\n",
       "      <td>53</td>\n",
       "      <td>30</td>\n",
       "      <td>27.9</td>\n",
       "      <td>7.500000e-01</td>\n",
       "      <td>59.6</td>\n",
       "      <td>8.900000e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1382440 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                qname  qstart  qend  qlength         sname  sstart  send  \\\n",
       "0        reprseq00001       1    86       86  reprseq00001       1    86   \n",
       "1        reprseq00001      30    44       86  reprseq14105      27    41   \n",
       "2        reprseq00002       1   125      125  reprseq00002       1   125   \n",
       "3        reprseq00002       1   125      125  reprseq13080       1   125   \n",
       "4        reprseq00002       1   125      125  reprseq17146      28   158   \n",
       "...               ...     ...   ...      ...           ...     ...   ...   \n",
       "1382435  reprseq34463      55    76      200  reprseq32916      39    60   \n",
       "1382436  reprseq34464       1   187      187  reprseq34464       1   187   \n",
       "1382437  reprseq34464       9   183      187  reprseq06995       1   181   \n",
       "1382438  reprseq34464       6    73      187  reprseq13037      48   109   \n",
       "1382439  reprseq34464      44    73      187  reprseq04060      14    43   \n",
       "\n",
       "         slength  pident  bitscore          eval   prob          pval  \n",
       "0             86     100     246.5  5.300000e-47  100.0  7.700000e-51  \n",
       "1             73      20      25.8  9.300000e-01   56.2  1.100000e-04  \n",
       "2            125     100     374.6  8.800000e-65  100.0  1.300000e-68  \n",
       "3            138      40     364.7  5.900000e-62  100.0  8.200000e-66  \n",
       "4            158      33     367.2  3.400000e-61  100.0  4.600000e-65  \n",
       "...          ...     ...       ...           ...    ...           ...  \n",
       "1382435      113      27      29.7  1.100000e+00   51.5  1.500000e-04  \n",
       "1382436      187     100     463.6  1.400000e-72  100.0  2.100000e-76  \n",
       "1382437      181      32     432.2  8.900000e-68  100.0  1.400000e-71  \n",
       "1382438      201      27      34.7  5.900000e-01   63.0  7.000000e-05  \n",
       "1382439       53      30      27.9  7.500000e-01   59.6  8.900000e-05  \n",
       "\n",
       "[1382440 rows x 13 columns]"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# repr-hits-pairwise #\n",
    "## define paths\n",
    "inter_dir = work_dir + 'intermediate/'\n",
    "all_by_all_output_dir = output_dir + 'prot-families/all-by-all/' + run_mode + '/'\n",
    "families_output_dir = output_dir + 'prot-families/families/'\n",
    "\n",
    "# load hhr_table\n",
    "table_hhr_filename = all_by_all_output_dir + 'table-hhr.txt'\n",
    "table_hhr = pd.read_csv(table_hhr_filename, sep=',')\n",
    "table_hhr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "530fd34c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qname</th>\n",
       "      <th>sname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>reprseq00001</td>\n",
       "      <td>reprseq14105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>reprseq00002</td>\n",
       "      <td>reprseq13080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>reprseq00002</td>\n",
       "      <td>reprseq17146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>reprseq00003</td>\n",
       "      <td>reprseq27929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>reprseq00004</td>\n",
       "      <td>reprseq08141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382434</th>\n",
       "      <td>reprseq34463</td>\n",
       "      <td>reprseq04091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382435</th>\n",
       "      <td>reprseq34463</td>\n",
       "      <td>reprseq32916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382437</th>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>reprseq06995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382438</th>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>reprseq13037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382439</th>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>reprseq04060</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1257076 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                qname         sname\n",
       "1        reprseq00001  reprseq14105\n",
       "3        reprseq00002  reprseq13080\n",
       "4        reprseq00002  reprseq17146\n",
       "6        reprseq00003  reprseq27929\n",
       "8        reprseq00004  reprseq08141\n",
       "...               ...           ...\n",
       "1382434  reprseq34463  reprseq04091\n",
       "1382435  reprseq34463  reprseq32916\n",
       "1382437  reprseq34464  reprseq06995\n",
       "1382438  reprseq34464  reprseq13037\n",
       "1382439  reprseq34464  reprseq04060\n",
       "\n",
       "[1257076 rows x 2 columns]"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get unique pairs\n",
    "## get from table_hhr qname-sname unique pairs of ids\n",
    "\n",
    "pair_table = table_hhr[['qname', 'sname']]\n",
    "pair_table = pair_table[pair_table['qname'] != pair_table['sname']]\n",
    "pair_table = pair_table.drop_duplicates()\n",
    "pair_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "d305661f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#test#\n",
    "pair_table_test = pair_table[:1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "dfe767ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create script to run in parallel\n",
    "n_cores = 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "d0ff5a59",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!! pair_table_test !!#\n",
    "\n",
    "# divide pair table into sub-tables\n",
    "chunk_size = int(len(pair_table_test) / n_cores)\n",
    "pair_table_parts = []\n",
    "for i in range(n_cores-1):\n",
    "    pair_table_parts.append(pair_table_test[i*chunk_size:(i+1)*chunk_size])\n",
    "pair_table_parts.append(pair_table_test[(n_cores-1)*chunk_size:])\n",
    "\n",
    "parts_dir = work_dir + 'tmp/prot-families/pair_table_chunks/'\n",
    "for i, pair_table_part in enumerate(pair_table_parts):\n",
    "    pair_table_part.to_csv(parts_dir + 'pair-table-' + str(i) + '.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "c0be5937",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run in parallel python script to process sub-table\n",
    "script_filepath =  work_dir + 'tmp/prot-families/run-pairwise-hits.sh'\n",
    "prob_threshold = 50\n",
    "\n",
    "cmd = '#!/bin/bash\\n\\n'\n",
    "cmd += 'source ' + pipeline_env_path + '/bin/activate\\n'\n",
    "\n",
    "for i, pair_table_part in enumerate(pair_table_parts):\n",
    "    pair_table_path = parts_dir + 'pair-table-' + str(i) + '.csv'    \n",
    "    cmd += 'nohup python3 {}/phage-pipeline/lib_phage/run_hits_pairwise_single_table.py {} {} {} {} {} &\\n'.format(\n",
    "    pipeline_env_path, work_dir, run_mode, i, prob_threshold, lib_pp_path)\n",
    "\n",
    "with open(script_filepath, 'w') as file_sh:\n",
    "    file_sh.write(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dd68082",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "42d3d96c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qname</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>qlength</th>\n",
       "      <th>sname</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>slength</th>\n",
       "      <th>pident</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>eval</th>\n",
       "      <th>prob</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>reprseq00001</td>\n",
       "      <td>1</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>reprseq00001</td>\n",
       "      <td>1</td>\n",
       "      <td>86</td>\n",
       "      <td>86</td>\n",
       "      <td>100</td>\n",
       "      <td>246.5</td>\n",
       "      <td>5.300000e-47</td>\n",
       "      <td>100.0</td>\n",
       "      <td>7.700000e-51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>reprseq00001</td>\n",
       "      <td>30</td>\n",
       "      <td>44</td>\n",
       "      <td>86</td>\n",
       "      <td>reprseq14105</td>\n",
       "      <td>27</td>\n",
       "      <td>41</td>\n",
       "      <td>73</td>\n",
       "      <td>20</td>\n",
       "      <td>25.8</td>\n",
       "      <td>9.300000e-01</td>\n",
       "      <td>56.2</td>\n",
       "      <td>1.100000e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>reprseq00002</td>\n",
       "      <td>1</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>reprseq00002</td>\n",
       "      <td>1</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>100</td>\n",
       "      <td>374.6</td>\n",
       "      <td>8.800000e-65</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.300000e-68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>reprseq00002</td>\n",
       "      <td>1</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>reprseq13080</td>\n",
       "      <td>1</td>\n",
       "      <td>125</td>\n",
       "      <td>138</td>\n",
       "      <td>40</td>\n",
       "      <td>364.7</td>\n",
       "      <td>5.900000e-62</td>\n",
       "      <td>100.0</td>\n",
       "      <td>8.200000e-66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>reprseq00002</td>\n",
       "      <td>1</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>reprseq17146</td>\n",
       "      <td>28</td>\n",
       "      <td>158</td>\n",
       "      <td>158</td>\n",
       "      <td>33</td>\n",
       "      <td>367.2</td>\n",
       "      <td>3.400000e-61</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4.600000e-65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382435</th>\n",
       "      <td>reprseq34463</td>\n",
       "      <td>55</td>\n",
       "      <td>76</td>\n",
       "      <td>200</td>\n",
       "      <td>reprseq32916</td>\n",
       "      <td>39</td>\n",
       "      <td>60</td>\n",
       "      <td>113</td>\n",
       "      <td>27</td>\n",
       "      <td>29.7</td>\n",
       "      <td>1.100000e+00</td>\n",
       "      <td>51.5</td>\n",
       "      <td>1.500000e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382436</th>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>1</td>\n",
       "      <td>187</td>\n",
       "      <td>187</td>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>1</td>\n",
       "      <td>187</td>\n",
       "      <td>187</td>\n",
       "      <td>100</td>\n",
       "      <td>463.6</td>\n",
       "      <td>1.400000e-72</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.100000e-76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382437</th>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>9</td>\n",
       "      <td>183</td>\n",
       "      <td>187</td>\n",
       "      <td>reprseq06995</td>\n",
       "      <td>1</td>\n",
       "      <td>181</td>\n",
       "      <td>181</td>\n",
       "      <td>32</td>\n",
       "      <td>432.2</td>\n",
       "      <td>8.900000e-68</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.400000e-71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382438</th>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>6</td>\n",
       "      <td>73</td>\n",
       "      <td>187</td>\n",
       "      <td>reprseq13037</td>\n",
       "      <td>48</td>\n",
       "      <td>109</td>\n",
       "      <td>201</td>\n",
       "      <td>27</td>\n",
       "      <td>34.7</td>\n",
       "      <td>5.900000e-01</td>\n",
       "      <td>63.0</td>\n",
       "      <td>7.000000e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1382439</th>\n",
       "      <td>reprseq34464</td>\n",
       "      <td>44</td>\n",
       "      <td>73</td>\n",
       "      <td>187</td>\n",
       "      <td>reprseq04060</td>\n",
       "      <td>14</td>\n",
       "      <td>43</td>\n",
       "      <td>53</td>\n",
       "      <td>30</td>\n",
       "      <td>27.9</td>\n",
       "      <td>7.500000e-01</td>\n",
       "      <td>59.6</td>\n",
       "      <td>8.900000e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1382440 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                qname  qstart  qend  qlength         sname  sstart  send  \\\n",
       "0        reprseq00001       1    86       86  reprseq00001       1    86   \n",
       "1        reprseq00001      30    44       86  reprseq14105      27    41   \n",
       "2        reprseq00002       1   125      125  reprseq00002       1   125   \n",
       "3        reprseq00002       1   125      125  reprseq13080       1   125   \n",
       "4        reprseq00002       1   125      125  reprseq17146      28   158   \n",
       "...               ...     ...   ...      ...           ...     ...   ...   \n",
       "1382435  reprseq34463      55    76      200  reprseq32916      39    60   \n",
       "1382436  reprseq34464       1   187      187  reprseq34464       1   187   \n",
       "1382437  reprseq34464       9   183      187  reprseq06995       1   181   \n",
       "1382438  reprseq34464       6    73      187  reprseq13037      48   109   \n",
       "1382439  reprseq34464      44    73      187  reprseq04060      14    43   \n",
       "\n",
       "         slength  pident  bitscore          eval   prob          pval  \n",
       "0             86     100     246.5  5.300000e-47  100.0  7.700000e-51  \n",
       "1             73      20      25.8  9.300000e-01   56.2  1.100000e-04  \n",
       "2            125     100     374.6  8.800000e-65  100.0  1.300000e-68  \n",
       "3            138      40     364.7  5.900000e-62  100.0  8.200000e-66  \n",
       "4            158      33     367.2  3.400000e-61  100.0  4.600000e-65  \n",
       "...          ...     ...       ...           ...    ...           ...  \n",
       "1382435      113      27      29.7  1.100000e+00   51.5  1.500000e-04  \n",
       "1382436      187     100     463.6  1.400000e-72  100.0  2.100000e-76  \n",
       "1382437      181      32     432.2  8.900000e-68  100.0  1.400000e-71  \n",
       "1382438      201      27      34.7  5.900000e-01   63.0  7.000000e-05  \n",
       "1382439       53      30      27.9  7.500000e-01   59.6  8.900000e-05  \n",
       "\n",
       "[1382440 rows x 13 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# there are some paths that are used further in pipeline #\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "## define paths\n",
    "output_dir = work_dir + 'output/'\n",
    "prot_families_dir = output_dir + 'prot-families/'\n",
    "prot_families_temp_dir = work_dir + 'tmp/prot-families/'\n",
    "\n",
    "inter_dir = work_dir + 'intermediate/'\n",
    "all_by_all_output_dir = output_dir + 'prot-families/all-by-all/' + run_mode + '/'\n",
    "families_output_dir = output_dir + 'prot-families/families/'\n",
    "\n",
    "## load representative mappings\n",
    "repr_dir = work_dir + 'output/prot-families/representative/'\n",
    "repr_table_filename = repr_dir + 'name-table.txt'\n",
    "repr_table = pd.read_csv(repr_table_filename, sep=',')\n",
    "nr = len(repr_table)\n",
    "nrd = len(str(nr))\n",
    "\n",
    "table_hhr_filename = all_by_all_output_dir + 'table-hhr.txt'\n",
    "table_hhr = pd.read_csv(table_hhr_filename, sep=',')\n",
    "table_hhr"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
